### ISSUE TITLE
New lint suggestion: `iter_copied_too_late`

### ISSUE THREAD

**Issue by workingjubilee** [rust-lang/rust-clippy#10172]
### What it does

This is the proposal of a Clippy `perf` lint to address the performance issue cited here: https://github.com/rust-lang/rust/issues/106539

It suggests moving `.copied()` in an iterator chain earlier if the resulting type would be a value type `T` anyways instead of `&T`.

This may result in confusing advice since it contradicts `iter_overeager_cloned`. However, to clone vs. copying things is genuinely different, so more experienced Rust users won't be too surprised.

### Lint Name

iter_copied_too_late

### Category

perf

### Advantage

Especially with simple scalar types, this results in the usage of vector instructions when lowered to LLVM IR.

### Drawbacks

Some false positives:
- if copying only removed a layer of indirection, from `&&T` to `&T` instead of `&T` to `T`
- type is too complex to vectorize (safest if it is a primitive or newtype of one)
- random other niche cases where this doesn't actually improve performance can happen also

### Example

```rust
pub fn scalarized_max<T: Ord + Copy>(a: &[T]) -> Option<T> {
    a.iter().max().copied()
}
```

This will typically drop to scalar code because LLVM can't immediately see that it can just dereference everything and reason about simple scalar types instaed. It could be written as:

```rust
pub fn vectorized_max<T: Ord + Copy>(a: &[T]) -> Option<T> {
    a.iter().copied().max()
}
```

**Comment by scottmcm** [comment]
> This may result in confusing advice since it contradicts `iter_overeager_cloned`. However, to clone vs. copying things is genuinely different, so more experienced Rust users won't be too surprised.

This might be useful, actually -- if the copy is expensive, they can switch to `.cloned()`, which would then be at the end.

It's a great additional justification for `.copied()` and `.cloned()` both existing.
